src/cost_estimates/fee_medians.rs

```diff
@@ -0,0 +1,348 @@
+                   "new_estimate_middle" => next_estimate.middle,
+                   "new_estimate_low" => next_estimate.low);
+        }
+    }
+}
+
+impl<M: CostMetric> FeeEstimator for WeightedMedianFeeRateEstimator<M> {
+    fn notify_block(
+        &mut self,
+        receipt: &StacksEpochReceipt,
+        block_limit: &ExecutionCost,
+    ) -> Result<(), EstimatorError> {
+        // Calculate sorted fee rate for each transaction in the block.
+        let mut working_fee_rates: Vec<FeeRateAndWeight> = receipt
+            .tx_receipts
+            .iter()
+            .filter_map(|tx_receipt| {
+                fee_rate_and_weight_from_receipt(&self.metric, &tx_receipt, block_limit)
+            })
+            .collect();
+
+        // If necessary, add the "minimum" fee rate to fill the block.
+        maybe_add_minimum_fee_rate(&mut working_fee_rates, self.full_block_weight);
+
+        // If fee rates non-empty, then compute an update.
+        if working_fee_rates.len() > 0 {
+            // Values must be sorted.
+            working_fee_rates.sort_by(|a, b| {
+                a.fee_rate
+                    .partial_cmp(&b.fee_rate)
+                    .unwrap_or(Ordering::Equal)
+            });
+
+            // Compute the estimate and update.
+            let block_estimate = fee_rate_estimate_from_sorted_weighted_fees(&working_fee_rates);
+            self.update_estimate(block_estimate);
+        }
+
+        Ok(())
+    }
+
+    fn get_rate_estimates(&self) -> Result<FeeRateEstimate, EstimatorError> {
+        Self::get_rate_estimates_from_sql(&self.db, self.window_size)
+    }
+}
+
+/// Computes a `FeeRateEstimate` based on `sorted_fee_rates` using a "weighted percentile" method
+/// described in https://en.wikipedia.org/wiki/Percentile#Weighted_percentile
+///
+/// The percentiles computed are [0.05, 0.5, 0.95].
+///
+/// `sorted_fee_rates` must be non-empty.
+fn fee_rate_estimate_from_sorted_weighted_fees(
+    sorted_fee_rates: &Vec<FeeRateAndWeight>,
+) -> FeeRateEstimate {
+    if sorted_fee_rates.is_empty() {
+        panic!("`sorted_fee_rates` cannot be empty.");
+    }
+
+    let mut total_weight = 0u64;
+    for rate_and_weight in sorted_fee_rates {
+        total_weight += rate_and_weight.weight;
+    }
+    let mut cumulative_weight = 0u64;
+    let mut percentiles = Vec::new();
+    for rate_and_weight in sorted_fee_rates {
+        cumulative_weight += rate_and_weight.weight;
+        let percentile_n: f64 =
+            (cumulative_weight as f64 - rate_and_weight.weight as f64 / 2f64) / total_weight as f64;
+        percentiles.push(percentile_n);
+    }
+
+    let target_percentiles = vec![0.05, 0.5, 0.95];
+    let mut fees_index = 0; // index into `sorted_fee_rates`
+    let mut values_at_target_percentiles = Vec::new();
+    for target_percentile in target_percentiles {
+        while fees_index < percentiles.len() && percentiles[fees_index] < target_percentile {
+            fees_index += 1;
+        }
```

Comment:
What happens when `percentiles[fees_index] == target_percentile`?

I think in the correct operation of the weighted percentile algorithm, `value` should equal `sorted_fee_rates[fee_index].fee_rate` in that case. Is that the case in this implementation?

Reply:
I exposed `fee_rate_estimate_from_sorted_weighted_fees` for test, and added test cases in `src/cost_estimates/tests/fee_medians.rs`.

@kantai I just put the answers from the program in to the right-hand side of the "expect".. What do you think?